node/subsystem/src/util.rs

```diff
@@ -0,0 +1,563 @@
+	/// task to handle those messages.
+	// TODO: the API here is likely not precisely what we want; figure it out more
+	// once we're implementing a subsystem which actually needs this feature.
+	// In particular, we're quite likely to want this to return a future instead of
+	// interrupting the active thread for the duration of the handler.
+	fn handle_unhashed_msg(_msg: Self::ToJob) -> Result<(), Self::Error> {
+		Ok(())
+	}
+}
+
+pub struct Jobs<Spawner, Job: JobTrait> {
+	spawner: Spawner,
+	running: HashMap<Hash, JobHandle<Job::ToJob>>,
+	outgoing_msgs: StreamUnordered<mpsc::Receiver<Job::FromJob>>,
+	job: std::marker::PhantomData<Job>,
+}
+
+impl<Spawner: Spawn, Job: JobTrait> Jobs<Spawner, Job> {
+	/// Create a new Jobs manager which handles spawning appropriate jobs.
+	pub fn new(spawner: Spawner) -> Self {
+		Self {
+			spawner,
+			running: HashMap::new(),
+			outgoing_msgs: StreamUnordered::new(),
+			job: std::marker::PhantomData,
+		}
+	}
+
+	/// Spawn a new job for this `parent_hash`, with whatever args are appropriate.
+	fn spawn_job(&mut self, parent_hash: Hash, run_args: Job::RunArgs) -> Result<(), Error> {
+		let (to_job_tx, to_job_rx) = mpsc::channel(JOB_CHANNEL_CAPACITY);
+		let (from_job_tx, from_job_rx) = mpsc::channel(JOB_CHANNEL_CAPACITY);
+		let (finished_tx, finished) = oneshot::channel();
+
+		let (future, abort_handle) = future::abortable(async move {
+			if let Err(e) = Job::run(parent_hash, run_args, to_job_rx, from_job_tx).await {
+				log::error!(
+					"{}({}) finished with an error {:?}",
+					Job::NAME,
+					parent_hash,
+					e,
+				);
+			}
+		});
+
+		// discard output
+		let future = async move {
+			let _ = future.await;
+			let _ = finished_tx.send(());
+		};
+		self.spawner.spawn(future)?;
+
+		// this handle lets us remove the appropriate receiver from self.outgoing_msgs
+		// when it's time to stop the job.
+		let outgoing_msgs_handle = self.outgoing_msgs.push(from_job_rx);
+
+		let handle = JobHandle {
+			abort_handle,
+			to_job: to_job_tx,
+			finished,
+			outgoing_msgs_handle,
+		};
+
+		self.running.insert(parent_hash, handle);
+
+		Ok(())
+	}
+
+	/// Stop the job associated with this `parent_hash`.
+	pub async fn stop_job(&mut self, parent_hash: Hash) -> Result<(), Error> {
+		match self.running.remove(&parent_hash) {
+			Some(handle) => {
+				Pin::new(&mut self.outgoing_msgs).remove(handle.outgoing_msgs_handle);
+				handle.stop().await;
+				Ok(())
+			}
+			None => Err(Error::JobNotFound(parent_hash)),
+		}
+	}
+
+	/// Send a message to the appropriate job for this `parent_hash`.
+	async fn send_msg(&mut self, parent_hash: Hash, msg: Job::ToJob) -> Result<(), Error> {
+		match self.running.get_mut(&parent_hash) {
+			Some(job) => job.send_msg(msg).await?,
+			None => return Err(Error::JobNotFound(parent_hash)),
+		}
+		Ok(())
+	}
+
+	/// Get the next message from any of the underlying jobs.
+	async fn next(&mut self) -> Option<Job::FromJob> {
```

Comment:
could we possibly `impl Stream`?

Reply:
yeah, i think it would be straightforward as we'd just forward all methods to `self.outgoing_msgs`

Reply:
It turns out to be a bit more complicated than that, but I got it done in 940c1112. Because this requires unsafe code, I'd appreciate particularly thorough review of that commit.